backends/ldap.go

```diff
@@ -0,0 +1,343 @@
+	}
+
+	if baseDN, ok := authOpts["ldap_base_dn"]; ok {
+		l.BaseDN = baseDN
+	} else {
+		ldapOk = false
+		missingOptions += " ldap_base_dn"
+	}
+
+	if groupBaseDN, ok := authOpts["ldap_group_base_dn"]; ok {
+		l.GroupBaseDN = groupBaseDN
+	}
+
+	if bindDN, ok := authOpts["ldap_bind_dn"]; ok {
+		l.BindDN = bindDN
+	} else {
+		ldapOk = false
+		missingOptions += " ldap_bind_dn"
+	}
+
+	if bindPass, ok := authOpts["ldap_bind_password"]; ok {
+		l.BindPass = bindPass
+	} else {
+		ldapOk = false
+		missingOptions += " ldap_bind_password"
+	}
+
+	if userFilter, ok := authOpts["ldap_user_filter"]; ok {
+		l.UserFilter = userFilter
+	} else {
+		ldapOk = false
+		missingOptions += " ldap_user_filter"
+	}
+
+	if groupFilter, ok := authOpts["ldap_group_filter"]; ok {
+		l.GroupFilter = groupFilter
+	}
+
+	if superuserFilter, ok := authOpts["ldap_superuser_filter"]; ok {
+		l.SuperuserFilter = superuserFilter
+	}
+
+	if aclTopicPatternAttribute, ok := authOpts["ldap_acl_topic_pattern_attribute"]; ok {
+		l.AclTopicPatternAttribute = aclTopicPatternAttribute
+	}
+
+	if aclAccAttribute, ok := authOpts["ldap_acl_acc_attribute"]; ok {
+		l.AclAccAttribute = aclAccAttribute
+	}
+
+	//Exit if any mandatory option is missing.
+	if !ldapOk {
+		return l, errors.Errorf("LDAP backend error: missing options:%s", missingOptions)
+	}
+
+	//Check if the LDAP server is reachable
+	ldapClient, err := l.LDAPClientFactory(l)
+	if err != nil {
+		log.Debugf("LDAP connection error: %s", err)
+		return l, err
+	}
+	l.LDAPClient = ldapClient
+
+	err = l.LDAPClient.Bind(l.BindDN, l.BindPass)
+	if err != nil {
+		log.Debugf("LDAP bind error: %s", err)
+		return l, err
+	}
+
+	return l, nil
+}
+
+func (l LDAP) GetUser(username, password, clientid string) (bool, error) {
+
+	searchRequest := ldap.NewSearchRequest(
+		l.BaseDN,
+		ldap.ScopeWholeSubtree,
+		ldap.NeverDerefAliases,
+		0,
+		0,
+		false,
+		fmt.Sprintf(l.UserFilter, ldap.EscapeFilter(username)),
+		[]string{"dn"},
+		nil,
+	)
+
+	searchResult, err := l.LDAPClient.Search(searchRequest)
+	if err != nil {
+		log.Debugf("LDAP user search error: %s", err)
+		return false, err
+	}
+	if len(searchResult.Entries) != 1 {
+		log.Debugf("LDAP user search returned %d entries", len(searchResult.Entries))
+		return false, nil
+	}
+
+	userDN := searchResult.Entries[0].DN
+
+	userLdapClient, err := l.LDAPClientFactory(l)
+	if err != nil {
+		log.Debugf("LDAP user connection error: %s", err)
+		return false, err
+	}
+	defer func(ldapClient LDAPClient) {
+		err := ldapClient.Close()
+		if err != nil {
+			log.Errorf("LDAP user cleanup error: %s", err)
+		}
+	}(userLdapClient)
+
+	err = userLdapClient.Bind(userDN, password)
+	if err != nil {
+		log.Debugf("LDAP user bind error: %s", err)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (l LDAP) GetSuperuser(username string) (bool, error) {
+
+	//If there's no superuser filter, assume all privileges for all users.
+	if l.SuperuserFilter == "" {
+		return false, nil
+	}
```

Comment:
Maybe I'm reading this wrong, but isn't it returning the contrary that the comment says?
I.e. when there's no filter set, `GetSuperuser` will always return false.

Reply:
must have slipped in a refactoring